dsl/flatten.cc

```diff
@@ -0,0 +1,304 @@
+
+        // this vector contains all the possible RHS target locations that we might move to
+        vector<ast::ClassDef::RHS_store *> targets;
+        // 0 and 1 both go into the class itself
+        targets.emplace_back(&rhs);
+        targets.emplace_back(&rhs);
+        // 2 and up go into the to-be-created `class << self` blocks
+        for (auto &tgt : nestedBlocks) {
+            targets.emplace_back(&tgt);
+        }
+
+        // move everything to its appropriate target
+        for (auto &expr : exprs) {
+            if (auto methodDef = ast::cast_tree<ast::MethodDef>(expr.expr.get())) {
+                methodDef->setIsSelf(expr.staticLevel > 0);
+            }
+            targets[expr.staticLevel]->emplace_back(std::move(expr.expr));
+        }
+
+        // generate the nested `class << self` blocks as needed and add them to the class
+        for (auto &body : nestedBlocks) {
+            auto classDef =
+                ast::MK::Class(loc, loc,
+                               make_unique<ast::UnresolvedIdent>(core::Loc::none(), ast::UnresolvedIdent::Class,
+                                                                 core::Names::singleton()),
+                               {}, std::move(body), ast::ClassDefKind::Class);
+            rhs.emplace_back(std::move(classDef));
+        }
+
+        return rhs;
+    }
+
+public:
+    FlattenWalk() {
+        newMethodSet();
+    }
+    ~FlattenWalk() {
+        ENFORCE(methodScopes.empty());
+    }
+
+    unique_ptr<ast::ClassDef> preTransformClassDef(core::Context ctx, unique_ptr<ast::ClassDef> classDef) {
+        newMethodSet();
+        return classDef;
+    }
+
+    unique_ptr<ast::Expression> postTransformClassDef(core::Context ctx, unique_ptr<ast::ClassDef> classDef) {
+        classDef->rhs = addMethods(ctx, std::move(classDef->rhs), classDef->loc);
+        return classDef;
+    };
+
+    // We might want to move sends as well: either if they're method modifiers like `private` or `protected` or if
+    // they're `sig`s. If so, then we'll treat them like we treat methods on our method stack
+    unique_ptr<ast::Send> preTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        auto &methods = curMethodSet();
+        if (methods.stack.size() == 0) {
+            return send;
+        }
+
+        if (send->fun == core::Names::sig()) {
+            methods.stack.emplace_back(methods.methods.size(), 0, send.get());
+            methods.methods.emplace_back();
+        }
+
+        return send;
+    }
+
+    unique_ptr<ast::Expression> postTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        auto &methods = curMethodSet();
+
+        // if this isn't the thing on the stack, then we can ignore it
+        if (methods.stack.size() == 0 || methods.stack.back().target != send.get()) {
+            return send;
+        }
+
+        // we may not need to move this expression at all: check first
+        if (methods.stack.back().idx == -1) {
```

Comment:
Slight preference for using `optional<u4>` instead of `int` with `-1` as a special value, but that's non-blocking.

Reply:
+1

Reply:
I've removed the index entirely, so there's no special sentinel value any more.